Update HAL Sleep manager test to cope with STM32 LPTIM HW

LMESTM opened the pull request:
### Description
Update HAL Sleep manager test to cope with STM32 LPTIM HW

those changes are generic and should not create any regression to other targets, but are required for oncoming changes on STM targets (#10701) 

In particular and as kindly suggested by Przemec S. :
1.	Add setup/teardown handler’s for all cases. This disables sys-tick, so there should be no unexpected lp ticker interrupt scheduling.
2.	Modify setup/teardown handler’s: remove suspension of lp/us tickers, so they can count as this is required by test_lock_unlock_test_check test case.
3.	Use TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check()) after setting interrupt to cope with STM specific handling (CMPOK interrupt with deep-sleep locked). This performs wait only if needed and will not affect other targets which do not need extra wait.
4.	Move sleep_manager_lock_deep_sleep() after TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check())
5. Use TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check()) in
test_lock_unlock_test_check to let lower layers manage deep sleep.
<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->


### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ x] Test update
    [ ] Breaking change

### Reviewers
@jeromecoutant @mprse 
<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->

LMESTM committed:
Update HAL Sleep manager test to cope with STM32 LPTIM HW

In particular and as kindly suggested by Przemec S. :
1. Add setup/teardown handler’s for all cases. This disables sys-tick,
so there should be no unexpected lp ticker interrupt scheduling.
2. Modify setup/teardown handler’s: remove suspension of lp/us tickers,
so they can count as this is required by test_lock_unlock_test_check test
case.
3. Use TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check()) after
setting interrupt to cope with STM specific handling (CMPOK interrupt with
deep-sleep locked). This performs wait only if needed and will not affect
other targets which do not need extra wait.
4. Move sleep_manager_lock_deep_sleep() after TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check())
5. Use TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check()) in
test_lock_unlock_test_check to let lower layers manage deep sleep.

mprse commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -68,32 +88,19 @@ utest::v1::status_t testcase_setup(const Case *const source, const size_t index_
-#endif
-#endif
-    ticker_suspend(get_us_ticker_data());
+
```
I suggest leaving lp ticker wrapper suspension (`lp_ticker_wrapper_suspend();`) .

LMESTM replied:
fixed now

mprse reviewed (approved):
LGTM

LMESTM committed:
Fix compilation issue for targets without LPTICKER or USTICKER

LMESTM committed:
Remove unused variable

and avoid associated warning ...

fkjagodzinski commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -189,7 +224,7 @@ void test_lock_unlock_test_check()
     // * sleep_manager_can_deep_sleep() returns true,
     // * sleep_manager_can_deep_sleep_test_check() returns true instantly.
-    TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep());
+    TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check());
```
I assume this is to cover CMPOK interrupt, correct? Instead of this, could you add a `wraparound_lp_protect()` call above, please?

This test case confirms that `sleep_manager_can_deep_sleep_test_check()` can actually replace the original `sleep_manager_can_deep_sleep()` in some greentea tests. The *_test_check* has to be compared with the original.

https://github.com/ARMmbed/mbed-os/blob/9974899a6b1947f0ad4c729a3608b2d172d6eb28/platform/mbed_power_mgmt.h#L150-L159

LMESTM replied:
wraparound_lp_protect() will not ensure that CMPOK has fired. It will just check that the counter is not to close to the wrap around limit (e.g. 0xFFFF). An busy wait could maybe be used instead ?

mprse replied:
@fkjagodzinski good point. I will look into this tomorrow.

fkjagodzinski commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -68,31 +88,30 @@ utest::v1::status_t testcase_setup(const Case *const source, const size_t index_
     osKernelSuspend();
+
 #if DEVICE_LPTICKER
-    ticker_suspend(get_lp_ticker_data());
```
The `lp_ticker_wrapper_suspend()` should not be called without calling `ticker_suspend(get_lp_ticker_data())` first.

https://github.com/ARMmbed/mbed-os/blob/9974899a6b1947f0ad4c729a3608b2d172d6eb28/hal/mbed_lp_ticker_wrapper.h#L52-L61

LMESTM replied:
That was done following @mprse earlier comment to suspend the wrapper. But that the same time we have the:  "Modify setup/teardown handler’s: remove suspension of lp/us tickers, so they can count as this is required by test_lock_unlock_test_check test case." .. so I guess that we should go back to NOT suspending the lp_ticker_wrapper - @mprse : do you think that would be ok ?

mprse committed:
Further modifications in Sleep Manager test.

Changes:
- restore the original form of setup/teardown handlers,
- test_lock_unlock_test_check(): do not use common ticker layer (Timer, Timeout). Use only ticker HAL layer.
- Increase DEEP_SLEEP_TEST_CHECK_WAIT_DELTA_US delta.

mprse committed:
sleep_manager_can_deep_sleep_test_check(): do not use ticker common layer to count elapsed time.

mprse commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -68,32 +88,22 @@ utest::v1::status_t testcase_setup(const Case *const source, const size_t index_
 {
     // Suspend the RTOS kernel scheduler to prevent interference with duration of sleep.
     osKernelSuspend();
+
```
In the current version, only kernel is suspended. Common ticker layer is enabled and HAL ticker layer is reinitialized: `us_ticker_init();`, `lp_ticker_init();`. I think that this may cause undefined behavior. If common ticker layer is already initialized, then `lp_ticker_init()/us_ticker_init()` will disable ticker interrupt and may break the time tracing. Ticker features like `Timer` and `Timeout` assume that common ticker layer works fine are used in `test_lock_unlock_test_check()` test case.
I think that a safer method would be to completely cut off higher layers while testing HAL and rely only on the HAL ticker layer.

LMESTM replied:
Ok - aligned the PR with your proposal

LMESTM committed:
Increase DEEP_SLEEP_TEST_CHECK_WAIT_DELTA_US

now that sleep_manager_can_deep_sleep_test_check() is based on wait_ns

fkjagodzinski commented on TESTS/mbed_hal/sleep_manager/main.cpp:
```diff
@@ -175,53 +213,88 @@ void test_sleep_auto()
+        TEST_ASSERT_FALSE(sleep_manager_can_deep_sleep());
+        TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep_test_check());
+        stop = lp_ticker_read();
+        TEST_ASSERT_UINT64_WITHIN(DEEP_SLEEP_TEST_CHECK_WAIT_DELTA_US, DEEP_SLEEP_TEST_CHECK_WAIT_US / 2,
```
```suggestion
        TEST_ASSERT(diff_us(start, stop, p_ticker_info) < DEEP_SLEEP_TEST_CHECK_WAIT_US);
```
With the `DEEP_SLEEP_TEST_CHECK_WAIT_DELTA_US` increased, we now have `TEST_ASSERT_UINT64_WITHIN(delta=1000, expected=1000, actual=1000)` so this assertion needs to be changed.

LMESTM replied:
Could even be somthing like that, don't you think?
```
        TEST_ASSERT(diff_us(start, stop, p_ticker_info) >= DEEP_SLEEP_TEST_CHECK_WAIT_US/2);
        TEST_ASSERT(diff_us(start, stop, p_ticker_info) < DEEP_SLEEP_TEST_CHECK_WAIT_US);
```

LMESTM replied:
I updated PR as proposed

fkjagodzinski reviewed (commented):
Thanks for updates @mprse & @LMESTM. This looks good for me; there's just one minor thing to fix.

LMESTM committed:
Update test_lock_unlock_test_check() assertion to cope with new timings

With the DEEP_SLEEP_TEST_CHECK_WAIT_DELTA_US increased,
we now have TEST_ASSERT_UINT64_WITHIN(delta=1000, expected=1000, actual=1000)
so this assertion needed to be updated.

What we need is the deep sleep to be enabled after the programed interrupt
has fired and before a 2ms timeout expiration, which means >= 1000 and < 2000.

fkjagodzinski committed:
Tests: SleepManager: Fix test_check for NRF5X

Add missing `lp_ticker_clear_interrupt()` in the interrput handler used
in `test_lock_unlock_test_check()` test.
Remove redefined `us_to_ticks()`.
